topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,620 @@
+				topology.AddLayer2Link(p.Ctx.Graph, p.Groups[blockdevGroupName], node, linkMetadata)
+			}
+		}
+
+	}
+
+	for i := range blockdev.Children {
+		childNode := p.Ctx.Graph.LookupFirstNode(graph.Metadata{"Name": blockdev.Children[i].getName()})
+		if childNode != nil {
+			topology.AddLink(p.Ctx.Graph, childNode, node, "connected", nil)
+		}
+	}
+
+	p.blockdevMap[blockdev.Name] = blockdevInfo{
+		ID:   blockdev.Name,
+		Node: node,
+	}
+	return groupNode
+}
+
+// Prior to version 2.33 lsblk generated JSON that encodes all values
+// as strings.  Version 2.33 changed how integers and booleans are
+// encoded.  unmarshalWeakTypeJSON() will decode either version of the
+// JSON.
+func (p *ProbeHandler) unmarshalWeakTypeJSON(jsonBytes []byte) ([]BlockDevice, error) {
+	var decoder *mapstructure.Decoder
+	var err error
+	var result []BlockDevice
+
+	deviceMap := make(map[string]interface{})
+	err = json.Unmarshal([]byte(jsonBytes), &deviceMap)
+	config := &mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &result,
+	}
+	if decoder, err = mapstructure.NewDecoder(config); err != nil {
+		return nil, err
+	}
+	if err := decoder.Decode(deviceMap["blockdevices"]); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func (p *ProbeHandler) unregisterBlockdev(id string) {
+	p.Lock()
+	defer p.Unlock()
+
+	info, ok := p.blockdevMap[id]
+	if !ok {
+		return
+	}
+
+	p.Ctx.Graph.Lock()
+	if err := p.Ctx.Graph.DelNode(info.Node); err != nil {
+		p.Ctx.Graph.Unlock()
+		p.Ctx.Logger.Error(err)
+		return
+	}
+	p.Ctx.Graph.Unlock()
+
+	delete(p.blockdevMap, id)
+}
+
+func (p *ProbeHandler) connect() error {
+	var (
+		cmdOut []byte
+		err    error
+		intf   interface{}
+		result Devices
+	)
+
+	lsblkPath := p.Ctx.Config.GetString("agent.topology.blockdev.lsblk_path")
+	if cmdOut, err = exec.Command(lsblkPath, "-pO", "--json").Output(); err != nil {
+		return err
+	}
+
+	if err = json.Unmarshal([]byte(cmdOut[:]), &intf); err != nil {
+		return err
+	}
+
+	if err = mapstructure.WeakDecode(intf, &result); err != nil {
+		return err
+	}
+
+	// loop through the devices in the current map to make sure they haven't
+	// been removed
+	for _, current := range p.blockdevMap {
+		p.deleteIfRemoved(current, result.Blockdevices)
+	}
+
+	for i := range result.Blockdevices {
+		p.registerBlockdev(result.Blockdevices[i], "")
+	}
+
+	return nil
+}
+
+func (p *ProbeHandler) getID(blockdev BlockDevice) string {
+	if blockdev.WWN != "" {
+		return blockdev.WWN
+	}
+	if blockdev.Serial != "" {
+		return blockdev.Serial
+	}
+	return blockdev.Name
+}
+
+func (p *ProbeHandler) getName(blockdev BlockDevice) string {
+	if blockdev.Mountpoint != "" {
+		return blockdev.Mountpoint
+	}
+	return blockdev.Name
+}
+
+func newMetricsFromBlockdev(blockdevPath string) *topology.BlockdevMetric {
+	var (
+		cmdOut []byte
+		err    error
+		stats  sysstat
+	)
+	if cmdOut, err = exec.Command("/usr/bin/iostat", "-dx", "-o", "JSON", blockdevPath).Output(); err != nil {
```

Comment:
does it make sense to make it part of the config ?

Reply:
yes - done.